src/pybamm/solvers/solution.py

```diff
@@ -588,8 +627,38 @@ def process_casadi_var(self, var_pybamm, inputs, ys_shape):
         }
         inputs_MX = casadi.vertcat(*[p for p in inputs_MX_dict.values()])
         var_sym = var_pybamm.to_casadi(t_MX, y_MX, inputs=inputs_MX_dict)
-        var_casadi = casadi.Function("variable", [t_MX, y_MX, inputs_MX], [var_sym])
-        return var_casadi
+
+        opts = {
+            "cse": True,
+            "inputs_check": False,
+            "is_diff_in": [False, False, False],
+            "is_diff_out": [False],
+            "regularity_check": False,
+            "error_on_fail": False,
+            "enable_jacobian": False,
+        }
+
+        # Casadi has a bug where it does not correctly handle arrays with
+        # zeros padded at the beginning or end. To avoid this, we add and
+        # subtract a small number to the variable to reinforce the
+        # variable bounds.
+        epsilon = 5e-324
+        var_sym = (var_sym - epsilon) + epsilon
+
+        var_casadi = casadi.Function(
+            "variable",
+            [t_MX, y_MX, inputs_MX],
+            [var_sym],
+            opts,
+        )
+
+        # Some variables, like interpolants, cannot be expanded
+        try:
+            var_casadi_out = var_casadi.expand()
+        except RuntimeError:
```

Comment:
This is just catching casadi's possible error. I believe they don't have a way of checking if the expansion will work ahead of time, otherwise this would be an if statement

Reply:
My concern was more that other runtime errors could be caught by this

Reply:
Ah ok, I understand. Changed this to
```python
except RuntimeError as error:
    if "'eval_sx' not defined for" not in str(error):
        raise error  # pragma: no cover
    var_casadi_out = var_casadi
```